aquadoggo/src/db/stores/query.rs

```diff
@@ -0,0 +1,1703 @@
+
+        // We always query one more row than needed to find out if there's more data. This
+        // information aids the user during pagination
+        let has_next_page = if rows.len() as u64 > page_size {
+            // Remove that last row from final results if it exists
+            rows.pop();
+            true
+        } else {
+            false
+        };
+
+        // Calculate the total number of (filtered) documents in this query
+        let total_count = if args
+            .pagination
+            .fields
+            .contains(&PaginationField::TotalCount)
+        {
+            Some(self.count(schema, args, list).await?)
+        } else {
+            None
+        };
+
+        // Determine cursors for pagination by looking at beginning and end of results
+        let start_cursor = if args
+            .pagination
+            .fields
+            .contains(&PaginationField::StartCursor)
+        {
+            rows.first().map(|row| Self::row_to_cursor(row, list))
+        } else {
+            None
+        };
+
+        let end_cursor = if args.pagination.fields.contains(&PaginationField::EndCursor) {
+            rows.last().map(|row| Self::row_to_cursor(row, list))
+        } else {
+            None
+        };
+
+        let pagination_data = PaginationData {
+            total_count,
+            has_next_page,
+            // @TODO: Implement backwards pagination, see related issue:
+            // https://github.com/p2panda/aquadoggo/issues/325
+            has_previous_page: false,
+            start_cursor,
+            end_cursor,
+        };
+
+        // Finally convert everything into the right format
+        let documents = Self::convert_rows(rows, list, &application_fields, schema.id());
+
+        Ok((pagination_data, documents))
+    }
+
+    /// Query number of documents in filtered collection.
+    pub async fn count(
+        &self,
+        schema: &Schema,
+        args: &Query<DocumentCursor>,
+        list: Option<&RelationList>,
+    ) -> Result<u64, DocumentStorageError> {
+        let application_fields = args.select.application_fields();
+
+        let from = from_sql(list);
+        let where_ = where_sql(schema.id(), &application_fields, list);
+        let and_filters = where_filter_sql(&args.filter, schema);
+
+        let count_sql = format!(
+            r#"
+            SELECT
+                COUNT(documents.document_id)
+
+            FROM
+                {from}
+
+                JOIN operation_fields_v1
+                    ON
+                        document_view_fields.operation_id = operation_fields_v1.operation_id
+                        AND
+                            document_view_fields.name = operation_fields_v1.name
+
+            WHERE
+                {where_}
+                {and_filters}
+
+            -- Group application fields by name to make sure we get actual number of documents
+            GROUP BY operation_fields_v1.name
```

Comment:
Just trying to grok this.... In order to `COUNT` the total number of documents, avoiding duplicates which the rows bring in, we group by field name. This then means the `COUNT` happens over the rows grouped by name. 

Doesn't this still leave duplicates as if multiple fields are queried, then there will be duplicates as the field name is different?

Sure I'm just misunderstanding something....

I wonder if we could use `COUNT(DISTINCT documents.document_id)` though?

Reply:
Maybe relevant : https://github.com/p2panda/aquadoggo/issues/335

Reply:
Could be that this sneaked in through a Replace All mistake, think you're right :+1:

Reply:
I remember now: We can't just group by document_id since there might be duplicates when querying relation lists